rw-stream-sink/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
+// DEALINGS IN THE SOFTWARE.
+
+//! This crate provides the `RwStreamSink` type. It wraps around a `Stream + Sink` that produces
+//! byte arrays, and implements `AsyncRead` and `AsyncWrite`.
+
+extern crate bytes;
+extern crate futures;
+extern crate tokio_io;
+
+use std::cmp;
+use std::io::Error as IoError;
+use std::io::ErrorKind as IoErrorKind;
+use std::io::{Read, Write};
+use bytes::{Buf, IntoBuf};
+use futures::{Async, AsyncSink, Poll, Stream, Sink};
+use tokio_io::{AsyncRead, AsyncWrite};
+
+/// Wraps around a `Stream + Sink` whose items are buffers. Implements `AsyncRead` and `AsyncWrite`.
+pub struct RwStreamSink<S> where S: Stream, S::Item: IntoBuf {
+    inner: S,
+    current_item: Option<<S::Item as IntoBuf>::Buf>,
+}
+
+impl<S> RwStreamSink<S> where S: Stream, S::Item: IntoBuf {
+    /// Wraps around `inner`.
+    #[inline]
+    pub fn new(inner: S) -> RwStreamSink<S> {
+        RwStreamSink {
+            inner: inner,
+            current_item: None,
+        }
+    }
+}
+
+impl<S> Read for RwStreamSink<S>
+    where S: Stream<Error = IoError>,
+          S::Item: IntoBuf,
+{
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, IoError> {
+        let mut written = 0;
+
+        loop {
+            let need_new_item = if let Some(ref i) = self.current_item {
+                !i.has_remaining()
+            } else {
+                true
+            };
+
+            if need_new_item {
+                self.current_item = match self.inner.poll() {
+                    Ok(Async::Ready(i)) => i.map(|b| b.into_buf()),
+                    Ok(Async::NotReady) => {
+                        if written == 0 {
+                            return Err(IoError::new(IoErrorKind::WouldBlock, "stream not ready"));
+                        } else {
+                            return Ok(written);
+                        }
+                    },
+                    Err(err) => {
+                        if written == 0 {
+                            return Err(err);
+                        } else {
+                            return Ok(written);
+                        }
+                    },
+                };
+            }
+
+            let current_item = match self.current_item {
+                Some(ref mut i) => i,
+                None => return Ok(written),
+            };
+
+            let to_copy = cmp::min(buf.len() - written, current_item.remaining());
+            if to_copy == 0 {
+                return Ok(written);
+            }
+
+            current_item.by_ref().take(to_copy).copy_to_slice(&mut buf[written..(written+to_copy)]);
```

Comment:
I'm actually not sure what the situation is here, because the documentation for `Read::read` implies that only explicit reads are problematic. I don't think that's true though, because I definitely ran into failures before switching to the explicit pointer functions. I would use them just to be safe, since I don't know if it's explicitly specified anywhere.

Reply:
I just changed the code to zero out the output buffer so that this is not a problem.